File: leetcode/142-linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def detectCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head is None:
            return None

        slow, fast = head, head
        while True:
           if (fast.next is None or
               fast.next.next is None):
               return None
           slow = slow.next
           fast = fast.next.next
           if slow is fast:
               break
        slow = head
        while slow is not fast:
            slow = slow.next
            fast = fast.next
        return slow

File: leetcode/142-linked-list-cycle-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
 public:
  ListNode *detectCycle(ListNode *head) {
    if (!head) return nullptr;
    ListNode *slow, *fast;
    for (slow = head, fast = head; ;) {
      if (!fast->next || !fast->next->next) return nullptr;
      slow = slow->next;
      fast = fast->next->next;
      if (slow == fast) break;
    }
    for (slow = head; slow != fast;
         slow = slow->next, fast = fast->next) ;
    return slow;
  }
};
